### PR TITLE
[web] fix query editing

### LINKED ISSUE #4565: "PRETTIFY" data POST/GET with json (when modify)
Hi ... I was wondering if it is possible (when modifying POST / GET data of a request) to be able to modify how they appear "before pressing the change key".

In the sense, instead of "exhausting" to modify every single parameter through a single line !!! (in this way)

```
user=...&passwd=...&a=0&bb=&cc=&dd=&ddd=&xmodify=01000100&nn=&n1=&w3=..&a3=&fgg=&ss=..&a=..&asa=.., .....
```
```
user=...
passwd=...
a=0
bb=
cc=
dd=
ddd=
xmodify=01000100
nn=
n1=
w3=..
a3=
fgg=
ss=..a=..&asa=.., .....
```

### PR DESCRIPTION
**Pull request by yesmeck**
#### Description

As https://github.com/mitmproxy/mitmproxy/issues/4565#issuecomment-823912831 said, an empty body would show when editing a query string. This PR tries to fix the query editing logic. Also, fixed query replacement.

I added a new API for retrieving and updating query string

fix #4565

#### Checklist

 - [x] I have updated tests where applicable.
 - [x] I have added an entry to the CHANGELOG.

### CODE REVIEW

**Review by mhils — COMMENTED** [review]
Thanks for the PR and sorry for getting at this so late. :) Please see my question below. 😃

File: web/src/js/components/contentviews/HttpMessage.tsx [hunk 1]
```diff
@@ -32,7 +32,13 @@ export default function HttpMessage({flow, message}: HttpMessageProps) {
     } else {
         url = MessageUtils.getContentURL(flow, message, contentView, maxLines + 1);
     }
-    const content = useContent(url, message.contentHash);
+    let content = useContent(url, message.contentHash);
+    if (flow.request.method === "GET" && edit) {
```
[the remark this answers is not in the document]
Comment (yesmeck) [reply web/src/js/components/contentviews/HttpMessage.tsx]: Yes, you are right. I will try to find a more proper way to fix this issue.

[the remark this answers is not in the document]
Comment (yesmeck) [reply web/src/js/components/contentviews/HttpMessage.tsx]: Updated  and rebased.

**Comment by mhils** [comment]
Thanks. How come you updated this PR from using JS APIs to adding a new REST API endpoint for queries?

**Comment by yesmeck** [comment]
Do you mean where did I get this idea from, or how did I update this PR?

**Comment by mhils** [comment]
If I remember correctly, the initial version in this PR used JavaScript to extract the query from the path (I can't take a look at the code again because you force-pushed). Now there is a new REST endpoint instead. What motivated that change?

**Comment by yesmeck** [comment]
Before using JavaScript to extract the query from the path, I tried to make the `content.data` endpoint return queries. But soon, I realised that POST requests could also have queries that make `content.data` endpoint not sufficient to handle both bodies and queries. That is why I chose to parse queries from the client side. 

After you pointed out that my changes would break response editing, I found that the query replacement is also broken, so I plan to fix it as well. This requires changing the `content.data` endpoint. But as I said above, using `content.data` endpoint to handle both bodies and queries is a bit confusing for me. So I decided to create a new API endpoint for query replacement. Now that I have an endpoint for query replacement, I thought, why not get queries from the new endpoint.

**Pull request closed without merging**